### PR TITLE
Generate api endpoints from roxygen comments

### PR DESCRIPTION
**Pull request by richfitz**
This PR creates the initial structure for roxygen2 support. It allows writing roxygen comments like

```
#' @porcelain 
#'   POST /path => json(OutputSchema)
#'   query a :: numeric
#'   query b :: integer
f <- function(a, b) {
  # implementation
}
```

and theen when running roxygen on a package will automatically create code that can be included into the api with a new method `$handle_package()`

This is a fairly large PR because parsing text is horrible.  So about half of this is trying to understand the format above, even though it's not that bad.  Then we work within the framework roxygen presents (see [the extending roxygen vigneette](https://cran.r-project.org/web/packages/roxygen2/vignettes/extending.html)) generates appropriate endpoint while hopefully throwing useful errors when things go wrong.

Most of the porcelain interface is supported, trying to get this through a real API will likely expose missing bits, but see the inst/examples/add2 files for a very simple example

Particular things to focus on:

* Are we happy with the syntax above? It will be hard to change once we merge
* Are we happy with the method name `$handle_package()` it looks a bit odd to me, I named it by analogy to `$handle()` but am not sure it's the right name

### CODE REVIEW

File: vignettes/roxygen.Rmd [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+
+```
+#'   body data :: binary(application/zip)
+```
+
+would refer to a zip input.
+
+## Binding state
+
+Finally consider binding state into the API
```
Comment (r-ash) [inline vignettes/roxygen.Rmd]: Want to add some more detail to this bit?

File: vignettes/roxygen.Rmd [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+```
+
+**For a json body** it is likely that you will want to add a schema
+
+```r
+#' @porcelain
+#'   POST /path => json(OutputSchema)
+#'   query a :: numeric
+#'   query b :: integer
+#'   body data :: json(InputSchema)
```
Comment (r-ash) [inline vignettes/roxygen.Rmd:128]: This looks really good

File: vignettes/roxygen.Rmd [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+
+* `json`: `porcelain::porcelain_returning_json`
+* `binary`: `porcelain::porcelain_returning_binary`
+* `generic` `porcelain::porcelain_returning`
+
+If if no arguments are provided we perform a call with no arguments (e.g., `GET /path => json` would use `porcelain::porcelain_returning_json()`). If arguments are given, then we will attempt to auto-quote these.
+
+## Adding inputs
+
+The above function used query parameters to provide inputs to the target function; you can specify query, body and bound state this way (path parameters remain in the path, as before).
```
Comment (r-ash) [inline vignettes/roxygen.Rmd:93]: You have an example of path parameters above but I think would be useful to see this next to other input types too and in the same format as them. With an example of what the function should look like.

File: R/roxygen.R [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+  } else {
+    env <- getNamespace(package)
+  }
+  fn <- env[["__porcelain__"]]
+  if (is.null(fn)) {
+    pkg <- packageName(env)
+    if (is.null(pkg)) {
+      stop("No endpoints found: input is not a package name or namespace")
+    }
+    stop(sprintf("No endpoints found in package '%s'", pkg))
```
Comment (r-ash) [inline R/roxygen.R:78]: Maybe a reminder to run `devtools::document()` here?

Comment (r-ash) [reply R/roxygen.R:78]: If this is the error you would see if you forgot to generate docs? Though I guess that would only work the first time

File: R/roxygen.R [hunk 5]
```diff
@@ -0,0 +1,267 @@
+## The typical approach here would be to make roxygen2 a hard
+## dependency, import either the whole package or the three generics
+## (roxy_tag_parse, roclet_process and roclet_output) then register
+## the S3 methods with '@export' as normal.  However, this requires
+## that we have roxygen2 available at runtime which is pretty weird
+## given we only need it when running the package. From R 3.6.0 we can
```
Comment (r-ash) [inline R/roxygen.R]:
```suggestion
## given we only need it when documenting the package. From R 3.6.0 we can
```
?

**Review by r-ash — COMMENTED** [review]
This looks great, re method naming `handle_package` seems fine or maybe `handle_package_endpoints` might make it clearer what it does. Or something like `handle_generated_endpoints` `handle_roxygen_endpoints`.

File: R/roxygen.R (hunk 4 above)
Comment (richfitz) [reply R/roxygen.R:78]: Yeah, I think we'll end up needing some sort of general "check my setup" system, but would be useful to understand the failure modes first

File: vignettes/roxygen.Rmd (hunk 1 above)
Comment (richfitz) [reply vignettes/roxygen.Rmd]: um yes, that looks I forgot there

**Review by r-ash — APPROVED** [review]
Nice, `include_package_endpoint` is good

**Pull request merged**